### PR TITLE
Added fixed_vector

### PR DESCRIPTION
**Pull request by Daddelhai**
Added a vector with fixed capacity.
Fixes Issue #31 

Usage like a normal Vector.
✔ std::get<>() support
✔ forward iterator
✔ backward iterator
✔ insert, emplace, erase

### CODE REVIEW

**Comment by tilsche** [comment]
Open question: Howto / should we allow references as elements?

File: include/nitro/lang/fixed_vector.hpp [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+            return iterator(&storage_[size_]);
+        }
+
+        const typename fixed_vector<T>::reverse_iterator crend() const noexcept
+        {
+            return reverse_iterator(&storage_[-1]);
+        }
+
+        template <size_t I>
+        T& get() noexcept
```
Comment (bmario) [inline include/nitro/lang/fixed_vector.hpp:426]: there shouldn't be a get for this

File: include/nitro/lang/fixed_vector.hpp [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+        void erase(std::size_t key)
+        {
+            if (key >= size())
+                raise("Key does not exsist!");
+
+            while (key + 1 < size() && key + 1 < capacity())
+            {
+                storage_[key] = storage_[key + 1];
+                ++key;
+            }
```
Comment (bmario) [inline include/nitro/lang/fixed_vector.hpp:446]: use storage.erase() instead

File: include/nitro/lang/fixed_vector.hpp [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+            --size_;
+        }
+
+    private:
+        std::size_t size_ = 0;
+        std::size_t capacity_ = 0;
+
+        std::unique_ptr<T[]> storage_;
```
Comment (bmario) [inline include/nitro/lang/fixed_vector.hpp:454]: typically, this memeber would be called data_

File: include/nitro/lang/fixed_vector.hpp [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+                raise("Key does not exsist!");
+
+            while (key + 1 < size() && key + 1 < capacity())
+            {
+                storage_[key] = storage_[key + 1];
+                ++key;
+            }
+            --size_;
+        }
+
```
Comment (bmario) [inline include/nitro/lang/fixed_vector.hpp:449]: I'm missing a data() function

**Review by bmario — CHANGES_REQUESTED** [review]
The methods which take `depackage_func` seems strange. Remove that.

**Comment by Daddelhai** [comment]
Still no refs possible

File: tests/fixed_vector_test.cpp [hunk 5]
```diff
@@ -0,0 +1,131 @@
+#define CATCH_CONFIG_MAIN
+#include <catch2/catch.hpp>
+
+#include <nitro/lang/fixed_vector.hpp>
+
+TEST_CASE("fixed vector init", "[lang]")
+{
+    nitro::lang::fixed_vector<std::int64_t> v(10);
```
Comment (bmario) [inline tests/fixed_vector_test.cpp:8]: Use constexpr variables for literal constants

File: tests/fixed_vector_test.cpp [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+
+    v.emplace_back(5);
+    v.emplace_back(7);
+
+    REQUIRE(v.size() == 2);
+    REQUIRE(v[0] == 5);
+    REQUIRE(v.at(1) == 7);
+    REQUIRE(v.back() == 7);
```
Comment (bmario) [inline tests/fixed_vector_test.cpp:24]: add a check for front() as well

**Review by bmario — CHANGES_REQUESTED** [review]
Please look into `noexcept` and add it appropiately.

Regarding the tests, what is unique requirement of fixed vector compared to a standard vector? Add a test for that.

File: tests/fixed_vector_test.cpp (hunk 5 above)
Comment (Daddelhai) [reply tests/fixed_vector_test.cpp:8]: To make fixed_vector able for constexpr i need CppStd 20 (like the std::vector too)

Comment (bmario) [reply tests/fixed_vector_test.cpp:8]: Now I see what you were talking about with the constexpr. I may have sent you on a different track here. What my original intent was, was for you to replace the number literals in the tests with a const (or constexpr) variable.

In this particular case:

```suggestion
    constexpr int SIZE = 10;
    //....
    nitro::lang::fixed_vector<std::int64_t> v(SIZE);
    // ...
    REQUIRE(v.capacity() == SIZE);
```

File: tests/fixed_vector_test.cpp [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+    nitro::lang::fixed_vector<std::int64_t> v(SIZE, INIT_VALUES);
+
+    v[0] = 7;
+    v[2] = 9;
+    REQUIRE(v[0] == 7);
+    REQUIRE(v[2] == 9);
+}
+
+TEST_CASE("fixed vector at() access", "[lang]")
```
Comment (bmario) [inline tests/fixed_vector_test.cpp:97]: Please add two more tests. They should test the copying and move from one vector to another. Those test should test the state of BOTH the target and source vector after the respective operation.

File: include/nitro/lang/fixed_vector.hpp [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+        using value_type = T;
+        using pointer = value_type*;
+        using const_pointer = const value_type*;
+        using reference = value_type&;
+        using const_reference = const value_type&;
+        using size_type = std::size_t;
+        using iterator = value_type*;
+        using const_iterator = const value_type*;
+        using reverse_iterator = std::reverse_iterator<iterator>;
+        using const_reverse_iterator = std::reverse_iterator<const_iterator>;
```
Comment (bmario) [inline include/nitro/lang/fixed_vector.hpp:53]: With all these fancy typedefs now in place, you should replace the "raw types" in function signatures in this file, if applicable.

So, for example `std::size_t` should be `size_type`, and `value_type&` should be `reference`.

**Review by bmario — CHANGES_REQUESTED** [review]

**Review by bmario — APPROVED** [review]

**Pull request merged**